### PR TITLE
sql: introduce virtual sequences and use with SERIAL

### LINKED ISSUE #22607: sql: Hibernate tries to load SERIAL types into an int
Hibernate tries to load SERIAL types into an int, which fails if they are too large - testDirectIdPropertyAccess

We can likely use sequences to combat this but will need to investigate and test. 

Referenced in [16491](https://github.com/cockroachdb/cockroach/issues/16491)

### LINKED ISSUE #24062: sql: 32 bit SERIAL type
Cockroach's SERIAL type is currently 64 bits. With the new added SEQUENCE functionality it is possible to implement SERIAL as a postgres compatible 32 bit type. 

https://github.com/cockroachdb/cockroach/issues/18688

This might involve setting the current SERIAL type to a supported backwards compatibility type and introducing another type under the hood for new instances of the type. Caching functionality needs to be implemented for performance.

### LINKED ISSUE #26730: sql: sequences should be automatically added to a table for postgres compatibility
This came out of hibernate's test suite, the following tests fail:

- org.hibernate.engine.spi.ExtraStateTest shouldMaintainExtraStateWhenUsingIdentityIdGenerationStrategy
- org.hibernate.id.CreateDeleteTest createAndDeleteAnEntityInTheSameTransactionTest
- org.hibernate.id.FlushIdGenTest testPersistBeforeTransaction

Due to a sequence not existing.

```sql
create table ChineseTakeawayRestaurant (
  id  bigserial not null,
  gobelinStars int4 not null,
  primary key (id)
)
insert into ChineseTakeawayRestaurant (gobelinStars) 
  values 3

select currval('ChineseTakeawayRestaurant_id_seq')
```

Postgres automatically creates a sequence whenever a serial or big serial type is used. We don't do this.

### PR DESCRIPTION
**Pull request by knz**
Very good idea by @BramGruneir 

New features: virtual sequences (opt-in) and create-sequence-upon-SERIAL (opt-in). See individual commits for details and release notes.

For reference:

- I made everything opt-in here (i.e. default to previous CockroachDB behavior) to minimize risk.
- UX with Hibernate will thus not automatically be improved. We need to inform users (and upgrade our tutorials/docs) that this experimental feature is available and how to use it to unblock problems.
- It is possible that making the new default be `virtual_sequence` will both be backward-compatible with performance and unlock UX for Hibernate users. This has @awoods187 preference I think. However I propose to only flip that switch after we've seen a week or two pass with this patch merged.

Fixes #22607 (using the new setting set to `sql_sequence`).
Fixes #24062 (using the new setting set to `sql_sequence`).
Fixes #26730 (using the new setting set to either `virtual_sequence` or `sql_sequence`).

### CODE REVIEW

**Comment by knz** [comment]
I tested locally and this seems to work well. Please have a look already. However I still need to iterate to iron out a couple of test failures.

**Comment by knz** [comment]
oh it so appears that there are no test failures. Wowza

**Comment by knz** [comment]
@mjibson since you plan to look at this let me flesh out the configuration options that andy W, Bram and I have discussed.

There would be a new cluster setting (name TBD) to serve as cluster-wide default.
Then there would be a new session var (name TBD) which, in every session, defaults to the cluster setting. This is the same configuration mechanism we're using for `distsql` and `experimental_opt` already.

The setting would have the following 3 values:

- `rowid`: SERIAL maps to `INT NOT NULL DEFAULT unique_rowid()`
- `virtual_sequence`: SERIAL maps to `INT NOT NULL DEFAULT nextval(seqname)` with a sequence created as per `CREATE SEQUENCE seqname  VIRTUAL`
- `sql_sequence`: SERIAL maps to sequences like PostgreSQL.

**Comment by knz** [comment]
We're thinking about having the cluster setting default to `virtual_sequence` for *new* clusters, `rowid` for *existing* clusters, with `sql_sequence` become opt-in for client apps that are still not happy with virtual sequences.

**Comment by mjibson** [comment]
Reading over the code, it looks like these new virtual sequences are a sugar wrapper over unique_rowid. The IMPORT stuff calls that same function, so I think the actual bytes generated by IMPORT are compatible with this. It appears that IMPORT just needs to be taught about SERIAL or something. Overall I think this will be an easy fix for IMPORT.

**Comment by mjibson** [comment]
I think we have two options here. The first and easiest one is to have IMPORT convert SERIAL like the `rowid` option you describe above to avoid the sequence descriptor problem completely. I'm completely ok with that for now.

The second option is a full solution. The current IMPORT PGDUMP code supports sequences https://github.com/cockroachdb/cockroach/blob/6618d68716ae5a74099843fe768e92df7d50ad19/pkg/ccl/importccl/read_import_pgdump.go#L230 by adding them to a list of descriptors in a resolver https://github.com/cockroachdb/cockroach/blob/6618d68716ae5a74099843fe768e92df7d50ad19/pkg/ccl/importccl/read_import_pgdump.go#L242 and then passing that resolver in https://github.com/cockroachdb/cockroach/blob/6618d68716ae5a74099843fe768e92df7d50ad19/pkg/ccl/importccl/read_import_pgdump.go#L252 and at https://github.com/cockroachdb/cockroach/blob/575ecbcfa0184b3a9cb2b1b3e5f3c12e2bfa2de3/pkg/ccl/importccl/import_stmt.go#L197

I don't think it'd be too hard to do similar SERIAL processing in the IMPORT PGDUMP stuff.

A difficulty may arise when processing CSV, though, since it's never had to make more than 1 descriptor and so lacks some of the plumbing. I'm looking at CSV now to see what would be required for this.

**Comment by knz** [comment]
Configuration option is ready and I extended the commit message to explain how it works.

**Comment by knz** [comment]
Discussed with @dt @mjibson : ccl import will use the `rowid` strategy always for now. Fixing

**Comment by knz** [comment]
New feature has tests now too.

**Comment by knz** [comment]
@mjibson I modified `importccl/import_stmt.go` as we discussed. It would be nice if you could review just that.

Also there's another thing. PostgreSQL has the SERIAL pseudo-type always imply NOT NULL (we didn't do that, incorrectly, previously). This patch changes this, so I had to modify the import CSV test. Please have a look at that too. Thanks

File: pkg/ccl/importccl/import_stmt.go [hunk 1]
```diff
@@ -191,6 +211,9 @@ func MakeSimpleTableDescriptor(
 		Sequence: &importSequenceOperators{},
 	}
 	affected := make(map[sqlbase.ID]*sqlbase.TableDescriptor)
+
+	log.Warningf(ctx, "WOO %q", create.String())
```
Comment (mjibson) [inline pkg/ccl/importccl/import_stmt.go]: Remove.

File: pkg/ccl/importccl/csv_internal_test.go [hunk 2]
```diff
@@ -46,11 +46,14 @@ func TestMakeSimpleTableDescriptorErrors(t *testing.T) {
 			stmt:  "create table a (i int, constraint a foreign key (i) references c (id))",
 			error: `this IMPORT format does not support foreign keys`,
 		},
+		{
```
Comment (mjibson) [inline pkg/ccl/importccl/csv_internal_test.go]: Is this test still accurate? I think this and the next test should be reverted.

File: pkg/ccl/importccl/import_stmt.go [hunk 3]
```diff
@@ -168,15 +177,26 @@ func MakeSimpleTableDescriptor(
...
 				continue
 			}
-			n := tree.MakeTableName("", tree.Name(def.Table.TableNameReference.String()))
-			def.Table.TableNameReference = &n
+			// Strip the schema/db prefix.
+			refTable, err := def.Table.Normalize()
+			if err != nil {
+				return nil, err
+			}
+			*refTable = tree.MakeUnqualifiedTableName(refTable.TableName)
```
Comment (mjibson) [inline pkg/ccl/importccl/import_stmt.go]: Thanks for fixing this.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28575)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/importccl/import_stmt.go, line 215 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LJyaZGPTbYeE6z8Z8KR-r8-215:-LK3QC8G5h_C4DcrJqsM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/ccl/importccl/import_stmt.go#L215)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Remove.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/csv_internal_test.go, line 49 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LJyoU6R6wbcRrxN-izc-r8-49:-LK3QCzoCeazUj8CczdP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/ccl/importccl/csv_internal_test.go#L49)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Is this test still accurate? I think this and the next test should be reverted.
</blockquote></details>

Done.

**Review by madelynnblue — APPROVED** [review]
CCL stuff LGTM.

**Review by BramGruneir — APPROVED** [review]
:lgtm:

Just a few nits and comments.  

> Reviewed 8 of 11 files at r2, 1 of 1 files at r3, 20 of 20 files at r4, 9 of 29 files at r5, 8 of 11 files at r6, 1 of 1 files at r7, 27 of 27 files at r8, 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-:-LK3UUeu0_rqCRI5aYWE:b-xplgrv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/serial.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3CdLK4mkAOI8xFSyY:-LK3CdLK4mkAOI8xFSyZ:b-18ewn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/serial.go#L116)):*
> ```Go
> 		return nil, nil, nil, err
> 	}
> 	if res != nil {
> ```

Can you add the first attempt to the loop to?

Might make sense to pull it out into a function.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 833 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3CqkeAuq2eETpPYBH:-LK3Cqkf6vP8Q4kqrmiz:b592e65) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/logictest/testdata/logic_test/sequences#L833)):*
> ```
> user root
> 
> # Test virtual sequences.
> ```

subtest please
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 88 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3OhUOBSzck5Rckvav:-LK3OhUOBSzck5Rckvaw:bkdwy9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L88)):*
> ```
> DROP TABLE serials; DROP TABLE smallbig; DROP TABLE serial
> ```

lint: extra lines
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 91 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3PUqCBePxbOcqBbQ4:-LK3PUqD5WTuCRTsC9uM:b-wavvsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L91)):*
> ```
> 
> 
> subtest serial_virtual_sequence
> ```

Please add a test that has to loop 2 times to get a sequence name.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 182 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3OzEU6R1KlXqHdWvQ:-LK3OzEU6R1KlXqHdWvR:bs4in85) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L182)):*
> ```
> 
> statement ok
> DROP TABLE serials; DROP TABLE smallbig; DROP TABLE serial
> ```

this can be a single drop table statement
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 276 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3POsT3e5TdErugbk3:-LK3POsT3e5TdErugbk4:but1koe) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L276)):*
> ```
> 
> statement ok
> DROP TABLE serials; DROP TABLE smallbig; DROP TABLE serial
> ```

can be a single drop table
___
*[pkg/sql/sqlbase/structured.proto, line 699 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3Lxc1AyCmBQLHQgu5:-LK3Lxc1AyCmBQLHQgu6:b-wolqz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/sqlbase/structured.proto#L699)):*
> ```protobuf
>     // Start value of the sequence.
>     optional int64 start = 4 [(gogoproto.nullable) = false];
> 	// Whether the sequence is virtual.
> ```

lint: tabs

**Comment by knz** [comment]
All tests pass :)

Let me clean this up for you then we'll let the tires kick the (master) road.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28575)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/serial.go, line 116 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3CdLK4mkAOI8xFSyY:-LK3XbyN72TgG9RNckz8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/serial.go#L116)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you add the first attempt to the loop to?

Might make sense to pull it out into a function.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 833 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3CqkeAuq2eETpPYBH:-LK3WjWBAod1hkTFri-W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/logictest/testdata/logic_test/sequences#L833)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

subtest please
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 88 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3OhUOBSzck5Rckvav:-LK3V-zV0ooEdbXW5PZ7:b54ire7) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L88)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

lint: extra lines
</blockquote></details>

I liked the little boost to readability of the logic test file. Oh well. Done.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 91 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3PUqCBePxbOcqBbQ4:-LK3V2KFEH1T1HoD4EKf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L91)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please add a test that has to loop 2 times to get a sequence name.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 182 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3OzEU6R1KlXqHdWvQ:-LK3Vjlu3DFiUB63Zw_C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L182)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

this can be a single drop table statement
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/serial, line 276 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3POsT3e5TdErugbk3:-LK3VkW-7h-Eu4YYRb1q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/699d2b838ba63eefc876f5eff2ffd3260983cf96/pkg/sql/logictest/testdata/logic_test/serial#L276)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

can be a single drop table
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 699 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28575#-LK3Lxc1AyCmBQLHQgu5:-LK3Vxbb8xO4Ek5JdYm9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ed2ef788fcd19511d6d41305f8f4567e3e48c67/pkg/sql/sqlbase/structured.proto#L699)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

lint: tabs
</blockquote></details>

Done.

**Comment by vilterp** [comment]
Nice work; so speedy! It's not on by default, right? Have to use the `VIRTUAL` keyword? Does it create it automatically if you use a `SERIAL` type?

**Comment by knz** [comment]
@vilterp read the release notes :)

**Comment by jordanlewis** [comment]
I'm confused - what does that cluster/session setting really do? Does it affect runtime behavior of existing SERIAL datatypes? Or does it only affect what's created when you make a table with SERIAL?

**Comment by knz** [comment]
The setting only affects new tables created inside that session.

**Pull request merged**